### PR TITLE
feat: add support for associating a Menu with a WebFrameMain

### LINKED ISSUE #44445: Use Writing Tools in macOS 15.1
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Are there plans to support apple intelligence's [`Writing Tools`](https://developer.apple.com/documentation/updates/apple-intelligence/) in electron?

The option is available in Chromium (tested on v128). Digging through the source code I found [`render_view_context_menu_mac.mm`](https://github.com/chromium/chromium/blob/main/chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac.mm) and [`text_services_context_menu.mm`](https://github.com/chromium/chromium/blob/main/ui/menus/cocoa/text_services_context_menu.mm), but I'm honestly none the wiser how it's implemented 🫠

Image

### Proposed Solution

`Writing Tools` could be exposed as a context menu `role`, added to the menu only if it's available. 

Also need to expose [`isWritingToolsActive`](https://developer.apple.com/documentation/appkit/nstextview/4431696-iswritingtoolsactive/) for UI/UX reasons, but I guess that can be of a lower priority.

### Alternatives Considered

nil

### Additional Information

Requirements to use apple intelligence:

- macOS 15.1+
- M-series Mac
- [Enable apple intelligence in settings](https://support.apple.com/en-sg/guide/mac-help/mchl46361784/mac)

### PR DESCRIPTION
**Pull request by MarshallOfSound**
Another Beach Sam classic 🥳 🏖️ 

This allows certain OS level features to activate such as Writing Tools, Autofill.. and Services.

There appears to be a bug in macOS where the responder chain isn't traversed if the menu is not popped up using an event, as such we spoof a fake mouse event at the right coordinates in the right window and use that to open the menu.

Notes: Added support for Autofill, Writing Tools and Services macOS level menu items in context menus via the new `frame` option in `menu.popup`

fixes #44445

### CODE REVIEW

File: shell/browser/api/electron_api_menu_mac.mm [hunk 1]
```diff
@@ -155,8 +166,26 @@
...
+    auto* rwhvm = static_cast<content::RenderWidgetHostViewMac*>(
+        frame.value()
+            ->render_frame_host()
+            ->GetOutermostMainFrameOrEmbedder()
+            ->GetView());
+    RenderWidgetHostViewCocoa* cocoa_view = rwhvm->GetInProcessNSView();
+    view = cocoa_view;
+  }
+
+  NSEvent* dummy_event = [NSEvent mouseEventWithType:NSEventTypeRightMouseDown
```
Comment (deepak1556) [inline shell/browser/api/electron_api_menu_mac.mm:179]: Can this be replaced with call to `ui::EventForPositioningContextMenuRelativeToWindow` ?

File: shell/browser/api/electron_api_menu_mac.mm [hunk 2]
```diff
@@ -155,8 +166,26 @@
   // be done manually.
   base::mac::ScopedSendingEvent sendingEventScoper;
```
Comment (deepak1556) [inline shell/browser/api/electron_api_menu_mac.mm:167]: `ui::showContextMenu` seems to create `ScopedSendingEvent` and `ScopedAllowApplicationTasksInNativeNestedLoop` if `allow_nested_tasks = true`, can these be removed in favor of that ?

**Review by deepak1556 — COMMENTED** [review]
Should this resolve https://github.com/electron/electron/issues/44445 ?

File: shell/browser/api/electron_api_menu_mac.mm (hunk 1 above)
Comment (MarshallOfSound) [reply shell/browser/api/electron_api_menu_mac.mm:179]: Unfortunately not, my fun fact is I figured all this out and _then_ realized Chrome had a similar method. Unfortunately that method defaults to using `[NSApp currentEvent]` if it has the correct type, but we almost always need to make our own to ensure the coords match what the user provided, not the coords that the event happened to have last.

File: shell/browser/api/electron_api_menu_mac.mm (hunk 2 above)
Comment (MarshallOfSound) [reply shell/browser/api/electron_api_menu_mac.mm:167]: I opted for explicitly doing it on our side of the contract instead of relying on a boolean flag in Chromes internals 🤔 Not fussed either way, just wanted as small a lift as possible.

File: shell/browser/api/electron_api_menu_mac.mm (hunk 1 above)
Comment (deepak1556) [reply shell/browser/api/electron_api_menu_mac.mm:179]: > Unfortunately that method defaults to using [NSApp currentEvent] if it has the correct type

Ah missed that, thanks for clarifying!

File: shell/browser/api/electron_api_menu_mac.mm [hunk 3]
```diff
@@ -57,14 +62,19 @@
   if (!native_window)
     return;
 
+  base::WeakPtr<WebFrameMain> weak_frame;
+  if (frame) {
```
Comment (ckerr) [inline shell/browser/api/electron_api_menu_mac.mm:66]: This is slightly brittle, it's possible to give the optional a `nullptr` s.t. the optional is nonempty with a null pointer and this paragraph results in a nullptr dereference

Maybe replace `std::optional<WebFrameMain*>` with just a `WebFrameMain*` that may be nullptr?

Comment (MarshallOfSound) [reply shell/browser/api/electron_api_menu_mac.mm:66]: > Maybe replace std::optional<WebFrameMain*> with just a WebFrameMain* that may be nullptr?

I don't think the current gin converter will take `Undefined` and convert it to `nullptr` 🤔 Nor do I think it should.

Given the above I don't think it's possible in this case to have the optional wrap a `nullptr` as the converter would never put one there, but that isn't enforced by semantics so I do see the issue still.

Maybe the gin converter should go to `std::reference_wrapper<T>` instead of `T*` to indicate it's non-null nature, but thats a bit of a lift for the whole codebase

Comment (samuelmaddock) [reply shell/browser/api/electron_api_menu_mac.mm:66]: Maybe an alternative could be to pass the [`content::GlobalRenderFrameHostToken`](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/render_frame_host.h;l=238-240;drc=f4daadec4cf496926ed23bb014749f7df81f6221) and retrieve the RenderFrameHost when needed. The token is used internally by WebFrameMain to lookup the RenderFrameHost. It's also recommended by the docs.

```cpp
// Returns the GlobalRenderFrameHostToken for this frame. Embedders should
// store this instead of a raw RenderFrameHost pointer.
virtual GlobalRenderFrameHostToken GetGlobalFrameToken() const = 0;
```

Comment (MarshallOfSound) [reply shell/browser/api/electron_api_menu_mac.mm:66]: @samuelmaddock This value comes from JS, do we have existing precedent for a `frameToken` being passed around 🤔

Comment (samuelmaddock) [reply shell/browser/api/electron_api_menu_mac.mm:66]: Ah sorry, I misunderstood the initial discussion. The `GlobalRenderFrameHostToken` could be an alternative to passing the weak ptr. I'm okay with either though.

As for the argument converter, I believe you could instead accept a `content::RenderFrameHost*` and it'll be converted from a WebFrameMain for you.

https://github.com/electron/electron/blob/da8288106be98912bf58aba346e927d7395b1679/shell/common/gin_converters/frame_converter.cc#L37-L48

Comment (MarshallOfSound) [reply shell/browser/api/electron_api_menu_mac.mm]: I opted for a `foo && foo.value()` check to ensure no nullptr in this case, I think it would be good to consider a codebase wide change to something more declarative like `std::reference_wrapper` but that requires more thought / everything doing it so for now I'll just make this bit ok

**Comment by ckerr** [comment]
macOS gn check failing:

> ERROR at //electron/shell/browser/api/electron_api_menu_mac.mm:15:11: Can't include this header from here.
#include "content/browser/renderer_host/render_widget_host_view_mac.h"
          ^----------------------------------------------------------

File: docs/api/menu.md [hunk 4]
```diff
@@ -73,6 +73,8 @@ The `menu` object has the following instance methods:
 
 * `options` Object (optional)
   * `window` [BaseWindow](base-window.md) (optional) - Default is the focused window.
+  * `frame` [WebFrameMain](web-frame-main.md) (optional) - Provide the relevant frame
+    if you want certain OS level features such as Writing Tools on macOS to function correctly. Typically this should be `params.frame` from the `context-menu` event on a WebContents.
```
Comment (erickzhao) [inline docs/api/menu.md:77]: minor punctuation

```suggestion
    if you want certain OS-level features such as Writing Tools on macOS to function correctly. Typically, this should be `params.frame` from the `context-menu` event on a WebContents.
```

File: docs/api/menu.md [hunk 5]
```diff
@@ -73,6 +73,8 @@ The `menu` object has the following instance methods:
 
 * `options` Object (optional)
   * `window` [BaseWindow](base-window.md) (optional) - Default is the focused window.
+  * `frame` [WebFrameMain](web-frame-main.md) (optional) - Provide the relevant frame
```
Comment (erickzhao) [inline docs/api/menu.md:76]: Can there be a default value here in the same way that we provide the focused window for the `window` parameter above?

Comment (samuelmaddock) [reply docs/api/menu.md:76]: I think this could be nice to add so all apps can take advantage of this feature by default. Especially in standard Electron use cases with one BrowserWindow containing a single WebContents. It might be worth adding a mention in the breaking changes doc too.

**Comment by codebytere** [comment]
@MarshallOfSound this should close https://github.com/electron/electron/issues/44445, correct?

**Comment by MarshallOfSound** [comment]
> @MarshallOfSound this should close https://github.com/electron/electron/issues/44445, correct?

Yeah I think so

File: docs/api/menu.md (hunk 5 above)
Comment (clavin) [reply docs/api/menu.md:76]: I made the default `options.window.webContents.mainFrame` when possible, as that seems to be a sensible default.

Comment (MarshallOfSound) [reply docs/api/menu.md:76]: Oh I specifically didn't want to enable this by default because it's super tricky to guess the right frame. Especially in a window with WebContentsViews or iframes and such. This would also magically opt people into these features which they may not want 🤔

**Comment by clavin** [comment]
Had to pause work on this. The feature works in regular use, however there seem to be some edge cases that our automated testing is catching. The current test failures boil down to a reentrancy bug where the context menu begins a new run loop that tries to open another context menu, which is not allowed by Chromium's `ui::ShowContextMenu` implementation.

File: docs/api/menu.md (hunk 5 above)
Comment (samuelmaddock) [reply docs/api/menu.md:76]: Here's one more idea to make getting a default frame easier:
`webContents.getFocusedWebContents()?.focusedFrame`

I've added `focusedFrame` in https://github.com/electron/electron/pull/45667

**Comment by clavin** [comment]
Update: [That approach worked](https://github.com/electron/electron/pull/45138/commits/e7cadf9e044b383d0fc69f689f0de199383ccfc7)! 😄

**Comment by KarthikeyanKanniappan** [comment]
As a curious query, so now would we have a role for writing tool which could be used in the context menu (on right click menu)
@clavin

File: shell/browser/api/electron_api_menu_mac.mm [hunk 6]
```diff
@@ -145,18 +156,27 @@
...
-  // setting flags in -[CrApplication sendEvent:], but since
-  // web-content menus are initiated by IPC message the setup has to
-  // be done manually.
-  base::mac::ScopedSendingEvent sendingEventScoper;
-
-  // Don't emit unresponsive event when showing menu.
-  [menu popUpMenuPositioningItem:item atLocation:position inView:view];
+
+  if (frame) {
+    auto* rfh = frame->render_frame_host()->GetOutermostMainFrameOrEmbedder();
```
Comment (samuelmaddock) [inline shell/browser/api/electron_api_menu_mac.mm:161]: Probably worth checking this since it can be a nullptr
```suggestion
    auto* frame_rfh = frame->render_frame_host();
    auto* rfh = frame_rfh ? frame_rfh->GetOutermostMainFrameOrEmbedder() : nullptr;
```

**Review by samuelmaddock — APPROVED** [review]
API LGTM

**Review by erickzhao — APPROVED** [review]
API LGTM

**Comment by craftzdog** [comment]
Awesome work!!!

Is it possible to display the "Speech" menu as well? It's usually available on regular macOS apps.
Screenshot 2025-06-02 at 22 03 14
